src/coreclr/jit/importer.cpp

```diff
@@ -7210,7 +7208,19 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                 // The array helper takes a native int for array length.
                 // So if we have an int, explicitly extend it to be a native int.
                 index = impImplicitIorI4Cast(index, TYP_I_IMPL);
-                op1   = gtNewHelperCallNode(CORINFO_HELP_LDELEMA_REF, TYP_BYREF, arr, index, type);
+
+                if ((gtGetArrayElementClassHandle(arr) == ldelemClsHnd) && impIsClassExact(ldelemClsHnd))
+                {
+                    JITDUMP("\nldelema of T[] with T exact: skipping covariant check\n");
+                    assert(lclTyp == TYP_REF);
+                    block->bbFlags |= BBF_HAS_IDX_LEN;
+                    optMethodFlags |= OMF_HAS_ARRAYREF;
+                    op1 = gtNewArrayIndexAddr(arr, index, lclTyp, ldelemClsHnd);
```

Comment:
It looks like the standard pattern here is to `goto ARR_LD`.  It would require the pop to be split to impStackTop/impPopStack(2) but pick up the various cases there (unless you're intentionally skipping those here?)

Reply:
Thanks! Addressed